### DISCUSSION TITLE
[Query] How can I only query direct children of the queried node?

### DISCUSSION THREAD

**Discussion by hponde** [tree-sitter/tree-sitter#845] [Q&A] [answered]
I'm having a hard time figuring out how to do this with the query language. For example if I want to only match top level imports in a python file, I would like to match all the nodes that are `future_import_statement`, `import_statement` or `import_from_statement` but only the ones that are direct children of the root node. So that when I do `queryObj.matches(rootNode)` I am able to get those directly without filtering it out later.

Is there an easy way to do that? For example I saw that there is an `#eq?` and tried: `((import_statement) @import) (#eq? @import.parent.parent null)`(which should capture the done calling it an `@import` and I tried to get only the root node direct children by checking for `.parent.parent` but that didn't work.

**Comment by hponde** [comment]
Since the root node was `module` in this case and there was no other `module` nodes in the tree, I was able to do the following: `(module (import_from_statement) @import)` and that worked well. But if there could be other `module` nodes in the tree then this wouldn't work as well.

**Comment by alemuller** [comment]
According to the python grammar, `module` is always at the root. In other words, each python file have a single `module` (this might not be true if working with injection).

**Reply by theHamsta** [reply]
But the queries are done per language. So without any additional information or predicates defined by the editor by the editor there is no way of finding out whether you're in a injected language or not. language injection does not create a common syntax tree.

**Reply by hponde** [reply]
I don't want to have to rely on that yeah, in case it breaks. If possible I'd like to only query nodes that are direct children of the node that I passed the query to, whatever it might be. It's more of a question of Query Syntax than anything else

**Comment by maxbrunsfeld** [comment, marked as answer]
This is an interesting use case that I hand't thought of, and I think it would be possible to add support for it. 

What I would probably do, is add a new concept of a "max depth" for a query cursor. Currently, you can already place bounds on the query by setting a byte range or a row/column range on the query cursor. This would extend that so you could bound the depth.

At the C API level, we'd add a method `ts_query_cursor_set_max_depth(TSQueryCursor *cursor, unsigned depth);`. If you passed in zero, it would mean that the depth is unlimited (which would be the default, on creating the query cursor).

I'm not sure about the WASM API. I think we'd need to add an additional optional argument to `Query.matches` and `Query.captures`.

**Reply by hponde** [reply]
I think this would work well for this case. But I would not want to add this unless it's useful in more cases, otherwise it could just bloat the API and not be useful for many?

I do not know about the underlying query/match logic. It is able to query for exact children already, so if we wanted to use that it seems we could have some special keyword like `rootQueryNode` or something, that matches the passed in node. That would also solve my problem.

@maxbrunsfeld I think you have a better sense of what would be better for the tool.
